compiler/semtypes.nim

```diff
@@ -456,8 +457,11 @@ proc semCaseBranch(c: PContext, t, branch: PNode, branchIndex: int,
     else:
       # constant sets and arrays are allowed:
       var r = semConstExpr(c, b)
-      # for ``{}`` we want to trigger the type mismatch in ``fitNode``:
-      if r.kind notin {nkCurly, nkBracket} or len(r) == 0:
+      if r.kind in {nkCurly, nkBracket} and len(r) == 0  and sonsLen(branch)==2:
```

Comment:
I don't the redundancy in these checks. Just move the 'len(r) == 0' check into the branch that deals with nkCurly, nkBracket.